rest/blip_api_test.go

```diff
@@ -3975,3 +3975,57 @@ func TestAttachmentWithErroneousRevPos(t *testing.T) {
 	assertStatus(t, resp, http.StatusOK)
 	assert.Equal(t, "goodbye cruel world", string(resp.BodyBytes()))
 }
+
+// CBG-2004: Test that prove attachment over Blip works correctly when receiving a ErrAttachmentNotFound
+func TestProveAttachmentNotFound(t *testing.T) {
+	rt := NewRestTester(t, &RestTesterConfig{
+		guestEnabled: true,
+	})
+	defer rt.Close()
+
+	bt, err := NewBlipTesterFromSpecWithRT(t, nil, rt)
+	assert.NoError(t, err, "Error creating BlipTester")
+	defer bt.Close()
+
+	attachmentData := []byte("attachmentA")
+	attachmentDataEncoded := base64.StdEncoding.EncodeToString(attachmentData)
+
+	bt.blipContext.HandlerForProfile[db.MessageProveAttachment] = func(msg *blip.Message) {
+		status, errMsg := base.ErrorAsHTTPStatus(db.ErrAttachmentNotFound)
+		msg.Response().SetError("HTTP", status, errMsg)
+	}
+
+	// Handler for when full attachment is requested
+	bt.blipContext.HandlerForProfile[db.MessageGetAttachment] = func(msg *blip.Message) {
+		resp := msg.Response()
+		resp.SetBody(attachmentData)
+		resp.SetCompressed(msg.Properties[db.BlipCompress] == "true")
+	}
+
+	// Initial set up
+	sent, _, _, err := bt.SendRev("doc1", "1-abc", []byte(`{"key": "val", "_attachments": {"attachment": {"data": "`+attachmentDataEncoded+`"}}}`), blip.Properties{})
+	require.True(t, sent)
+	require.NoError(t, err)
+
+	err = rt.WaitForPendingChanges()
+	require.NoError(t, err)
+
+	// Should log:
+	// "Peer sent prove attachment error 404 attachment not found, falling back to getAttachment for proof in doc <ud>doc1</ud> (digest sha1-wzp8ZyykdEuZ9GuqmxQ7XDrY7Co=)"
+	base.SetUpTestLogging(t, base.LevelInfo, base.KeyAll)
+
+	// Use different attachment name to bypass digest check in ForEachStubAttachment() which skips prove attachment code
+	// Set attachment to V2 so it can be retrieved by RT successfully
```

Comment:
Attachment cannot be retrieved from the bucket if the version is not included like so.

The GET attachment handler for `/{db}/{doc}/{attach}` assumes that the attachment version is version 1 (even though it's stored in the bucket as a v2 attachment) so it can't find it due to the name being incorrect. This is due to the attachment metadata not having a version tag which `GetAttachmentVersion` parses.

Possible bug, or could be Blip tester behaviour only.

Reply:
Still fails if not including attachment version when not hitting Prove Attachment code. If the version is included it also gets stripped out so assuming it is a blip tester problem or else this problem would have already come up.